### PR TITLE
[netease] IPs outside mainland China may access via CDNs.

### PR DESCRIPTION
**Pull request by jialongl**
## Please follow the guide below
- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your _pull request_ (like that [x])
- Use _Preview_ tab to see how your _pull request_ will actually look like

---
### Before submitting a _pull request_ make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
### What is the purpose of your _pull request_?
- [x] Bug fix
- [ ] New extractor
- [ ] New feature

---
### Description of your _pull request_ and other information

Like #7514, this commit makes downloading mp3 from netease possible again, by using hostnames that resolve to CNAMEs on CDN.

Should I revert 75af5d59ae5c4c6549cce5357a8a8d060c3f462c to enable the tests?

### CODE REVIEW

File: youtube_dl/extractor/neteasemusic.py [hunk 1]
```diff
@@ -48,8 +48,9 @@ def extract_formats(self, info):
             # from NetEase's CDN provider that can be used if m5.music.126.net does not
             # work, especially for users outside of Mainland China
             # via: https://github.com/JixunMoe/unblock-163/issues/3#issuecomment-163115880
-            for host in ('http://m5.music.126.net', 'http://115.231.74.139/m1.music.126.net',
-                         'http://124.40.233.182/m1.music.126.net', 'http://203.130.59.9/m1.music.126.net'):
+            # 2016-AUG-23: For users outside mainland China, use "(p|m)[0-9].music.126.net"
+            # Some of which resolve to CDN CNAMEs, such as "p.music.126.net.usgcac.cdnetworks.net.", "m3.music.126.net.wscdns.com."
+            for host in ('http://%c%d.music.126.net' % (c, i) for c in ('p', 'm') for i in xrange(1, 10)):
```
Comment (dstftw) [inline youtube_dl/extractor/neteasemusic.py]: There is no xrange in python 3.

**Comment by dstftw** [comment]
This does not seem to work at all:
```
PS C:\dev\youtube-dl\master> py -3.5 .\youtube_dl\__main__.py http://music.163.com/#/song?id=32102397 -F -v
[debug] System config: []
[debug] User config: []
[debug] Command-line args: ['http://music.163.com/#/song?id=32102397', '-F', '-v']
[debug] Encodings: locale cp1251, fs mbcs, out cp866, pref cp1251
[debug] youtube-dl version 2016.08.22
[debug] Git HEAD: 1b79c19
[debug] Python version 3.5.2 - Windows-10-10.0.10240-SP0
[debug] exe versions: ffmpeg N-81192-g04da20e, ffprobe N-81192-g04da20e, rtmpdump 2.4
[debug] Proxy map: {}
[netease:song] 32102397: Downloading song info
ERROR: No video formats found; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "C:\dev\youtube-dl\master\youtube_dl\YoutubeDL.py", line 691, in extract_info
    ie_result = ie.extract(url)
  File "C:\dev\youtube-dl\master\youtube_dl\extractor\common.py", line 347, in extract
    return self._real_extract(url)
  File "C:\dev\youtube-dl\master\youtube_dl\extractor\neteasemusic.py", line 165, in _real_extract
    self._sort_formats(formats)
  File "C:\dev\youtube-dl\master\youtube_dl\extractor\common.py", line 911, in _sort_formats
    raise ExtractorError('No video formats found')
youtube_dl.utils.ExtractorError: No video formats found; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
```

**Comment by jialongl** [comment]
@dstftw Thanks for your quick response. I'll fix `xrange`

Yea your URL fail on my end too.
But it seems that `http://music.163.com/#/song?id=32102397` is a piece of music behind paywall, thus netease's HTTP API returns me `None` for the `['songs'][0]['mp3Url']` key. I can't make magic from that.

My (smoke) test cases (done on zsh) were:
* http://music.163.com/#/album?id=3317645
* http://music.163.com/#/song?id=423406359

```
% ./youtube-dl -g -v 'http://music.163.com/#/album?id=3317645'
[debug] System config: []
[debug] User config: []
[debug] Command-line args: [u'-g', u'-v', u'http://music.163.com/#/album?id=3317645']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2016.08.22
[debug] Git HEAD: 7e5bfbd
[debug] Python version 2.7.12 - Darwin-15.6.0-x86_64-i386-64bit
[debug] exe versions: none
[debug] Proxy map: {}
http://p1.music.126.net/HbWL3hxnYMKsEwPqN1wWeg==/3400789465465085.mp3
http://p1.music.126.net/dtgUV2t3ffUTSBWYak1DNg==/3439272372597256.mp3
http://p1.music.126.net/kbHQQA3ngnCKkazNCoB6ww==/3412884093377037.mp3
http://p1.music.126.net/FyOC4qJVrJTWw034Xd1Jtw==/3400789465465086.mp3
http://p1.music.126.net/zJGxpm3OHPRnhR1w5kXjmQ==/3439272372597257.mp3
http://p1.music.126.net/LoWCjWsH2cqJbzFNwNz2Vw==/3412884093377038.mp3
http://p1.music.126.net/sVbWmLwhv_iokLSDsg14Bw==/3344714372967401.mp3
http://p1.music.126.net/hghicqYjXZ7fMBQxRGXTlA==/3417282140081837.mp3
http://p1.music.126.net/CnNt1BqUPVYqOg9d3quNVQ==/3412884093377039.mp3
http://p1.music.126.net/lLltn-q9_7t4R5z4vilZyA==/3344714372967402.mp3
http://p1.music.126.net/zLOTdB_UCy7ShUyGMg4ctw==/3424978721364014.mp3
http://p1.music.126.net/vGG0mj9m-LHUvj6tCUCsuw==/3439272372597259.mp3
```

```
0:14% ./youtube-dl -g -v 'http://music.163.com/#/song?id=423406359'
[debug] System config: []
[debug] User config: []
[debug] Command-line args: [u'-g', u'-v', u'http://music.163.com/#/song?id=423406359']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2016.08.22
[debug] Git HEAD: 7e5bfbd
[debug] Python version 2.7.12 - Darwin-15.6.0-x86_64-i386-64bit
[debug] exe versions: none
[debug] Proxy map: {}
http://p1.music.126.net/Sa6Os979UU5uCO6kCK0umg==/3265549572123018.mp3
```

And yes, I am actually able to download them.
Could you try some other free pieces?

Regards,
Jialong

File: youtube_dl/extractor/neteasemusic.py [hunk 2]
```diff
@@ -48,8 +48,9 @@ def extract_formats(self, info):
             # from NetEase's CDN provider that can be used if m5.music.126.net does not
             # work, especially for users outside of Mainland China
             # via: https://github.com/JixunMoe/unblock-163/issues/3#issuecomment-163115880
```
Comment (dstftw) [inline youtube_dl/extractor/neteasemusic.py]: This comment is now irrelevant.

**Comment by dstftw** [comment]
Squash commits when done.

**Comment by caizixian** [comment]
Per code of `you-get`, adding `'/'.join(info.get('mp3Url').split('/')[0:3])` and `"http://p{}".format(info.get('mp3Url').split('/')[2][1:])` to hosts list might help and it works on my side.

Need more feedbacks, so no new Pull Request atm.

**Comment by denobisipsis** [comment]
standard access:

- needed params & enseckey

1- generate random 16 bytes key from "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"

2- params:

	$params=base64_encode(encryptaes128(\''.$json.'\',"0CoJUm6Qyw8W8jud"));
	$params=urlencode(base64_encode(encryptaes128($params,"'.$key .'")));
        
       iv="0102030405060708"
      needed padding

2- enseckey is rsa:
     - modulus 00e0b509f6259df8642dbc35662901477df22677ec152b5ff68ace615bb7b725152b3ab17a876aea8a5aa76d2e417629ec4ee341f56135fccf695280104e0312ecbda92557c93870114af6c9d05c4f7f0c3685b7a46bee255932575cce10b424d813cfe4875d3e82047b97ddef52741d546b8e289dc6935b3ece0462db0a22b8e7
    -exponent starndard 010001

    -encrypt key

**Pull request closed without merging**